core/domain/translation_domain.py

```diff
@@ -18,9 +18,258 @@
+        self._translatable_contents = {}
+        self._register_all_translatable_fields()
+        return copy.deepcopy(self._translatable_contents)
+
+    def get_all_contents_which_needs_translations(
+        self,
+        entityTranslation: EntityTranslation
+    ) -> List[TranslatableContent]:
+        """Returns a list of TranslatableContent instances which needs
+        translation or which needs updates in an existing translations.
+
+        Args:
+            entityTranslation: EntityTranslation. An object storing existing
+                translation of an entity.
+
+        Returns:
+            list(TranslatableContent). Returns a list of TranslatableContents.
+        """
+        contents_which_need_translation = []
+        content_ids_for_translated_contents = (
+            entityTranslation.translations.keys())
+
+        for translatable_item in self.get_translatable_fields().values():
+            # translatable_item is of type TranslatableContent.
+            if translatable_item.content == '':
+                continue
+            if (
+                not translatable_item.content_id in
+                content_ids_for_translated_contents
+            ):
+                contents_which_need_translation.append(translatable_item)
+            elif (
+                entityTranslation.translations[
+                translatable_item.content_id].needs_update
+            ):
+                contents_which_need_translation.append(translatable_item)
+
+        return contents_which_need_translation
+
+
+class EntityTranslation:
+    """The EntityTranslation represents an EntityTransaltionsModel for a given
+    version of an entity in a given language.
+
+    Args:
+        entity_id: str. The id of the corresponding entity.
+        entity_type: str. The type of the corresponding entity.
+        entity_version: str. The version of the corresponding entity.
+        language_code: str. The language code for a corresponding entity.
+        translations: dict(str, TranslatedContent). The translated content.
+    """
+
+    def __init__(
+        self,
+        entity_id: str,
+        entity_type: str,
+        entity_version: int,
+        language_code: str,
+        translations: Dict[str, TranslatedContent]
+    ):
+        self.entity_id = entity_id
+        self.entity_type = entity_type
+        self.entity_version = entity_version
+        self.language_code = language_code
+        self.translations = translations
+
+
+class TranslatableContent:
+    """TranslatablesContent represents a content of a translatable object which
+    can be translated into multiple languages.
+
+    Args:
+        content_id: str. The id of the corresponding content.
+        content: str|list(str). The content which can be translated.
+        content_type: str. The type of the corresponding content.
+    """
+
+    def __init__(
+        self,
+        content_id: str,
+        content: str|List[str],
+        content_type: str
```

Comment:
The typing for this should be more restricted, right -- it can only be one of a few string literals/enums? Having this be a generic str doesn't reflect that. /cc @vojtechjelinek